app/main.py

```diff
@@ -1,34 +1,144 @@
+            if any(deck.is_alive for deck in ship.decks):
+                return "Hit!"
+            return "Sunk!"
+        return "Miss!"
+
+    def print_field(self) -> None:
+        field = [["~" for i in range(10)] for j in range(10)]
+        for ship in self.field.values():
+            for deck in ship.decks:
+                if not ship.is_drowned:
+                    if deck.is_alive:
+                        field[deck.coordinates[0]][
+                            deck.coordinates[1]] = "\u25A1"
+                    else:
+                        field[deck.coordinates[0]][deck.coordinates[1]] = "*"
+                    continue
+                field[deck.coordinates[0]][deck.coordinates[1]] = "x"
+        for row in field:
+            for column in row:
+                print(column, end=" ")
+            print()
+
+    def __validate_field(self) -> None:
+        data = set(self.__create_ships().values())
+        if len(self.ships) != 10:
+            raise Exception("The number of ships should be equal to 10")
+        self.__is_another_ship_in_neighbor_cell(data)
+        self.__correct_num_of_ships_with_definite_decks_num(data)
+
+    @staticmethod
+    def __correct_num_of_ships_with_definite_decks_num(
+            data: set[Ship]
+    ) -> None:
+        num_of_decs_and_ships = {1: 0, 2: 0, 3: 0, 4: 0}
+        for ship in data:
+            num_of_decks = len(ship.decks)
+            if num_of_decks in num_of_decs_and_ships:
+                num_of_decs_and_ships[num_of_decks] += 1
+                continue
+            raise Exception(
+                f"ship can't has {num_of_decks} decks,"
+                f" the value must be in range 1...4"
+            )
+        assert num_of_decs_and_ships == {1: 4, 2: 3, 3: 2, 4: 1}, (
+            "should be 4 single-deck ships; 3 double-deck ships;"
+            " 2 three-deck ships; 1 four-deck ship;"
+        )
+
+    def __is_another_ship_in_neighbor_cell(self, data: set[Ship]) -> None:
+        for ship in data:
+            for ship_deck in ship.decks:
+                for deck in self.field:
+                    ship_of_the_deck = self.field.get(deck)
+                    if (
+                            abs(ship_deck.coordinates[0] - deck[0]) <= 1
+                            and abs(ship_deck.coordinates[1] - deck[1]) <= 1
+                    ):
+                        all_decks_of_ship_of_the_deck = [
+                            deck.coordinates for deck in ship_of_the_deck.decks
+                        ]
+                        all_decks_of_ship = [
+                            deck.coordinates for deck in ship.decks
+                        ]
+                        if all_decks_of_ship_of_the_deck != all_decks_of_ship:
+                            raise Exception(
+                                "ships shouldn't be located"
+                                " in the neighboring cells"
+                            )
```

Comment:
Try to simplify this logic, so hard to read this
At least you can move the inner block to a function